From 3d8d0de8c7b0d479b4f72e446af946ae1fd2ac72 Mon Sep 17 00:00:00 2001
From: Maxime Leroy <maxime@leroys.fr>
Date: Fri, 5 Sep 2025 09:09:58 +0200
Subject: [PATCH 4/7] zebra: kernel get_speed take vrf_id and ifname as param

kernel_get_speed() doesn't need the full interface object; it only needs
the interface name and the VRF id to open the right socket/ioctl. Update
the prototype and callers accordingly.

This will be used in the next commit.

Signed-off-by: Maxime Leroy <maxime@leroys.fr>
---
 zebra/if_netlink.c | 14 +++-----------
 zebra/interface.c  |  4 ++--
 zebra/rt.h         |  2 +-
 zebra/rt_socket.c  |  6 ++++--
 4 files changed, 10 insertions(+), 16 deletions(-)

diff --git a/zebra/if_netlink.c b/zebra/if_netlink.c
index a214914171..c1ff967587 100644
--- a/zebra/if_netlink.c
+++ b/zebra/if_netlink.c
@@ -256,13 +256,12 @@ static void netlink_vrf_change(struct nlmsghdr *h, struct rtattr *tb,
 		ctx, *(uint32_t *)RTA_DATA(attr[IFLA_VRF_TABLE]));
 }
 
-static uint32_t get_iflink_speed(struct interface *interface, int *error)
+uint32_t kernel_get_speed(vrf_id_t vrf_id, const char *ifname, int *error)
 {
 	struct ifreq ifdata;
 	struct ethtool_cmd ecmd;
 	int sd;
 	int rc;
-	const char *ifname = interface->name;
 	uint32_t ret;
 
 	if (error)
@@ -280,8 +279,7 @@ static uint32_t get_iflink_speed(struct interface *interface, int *error)
 
 	/* use ioctl to get speed of an interface */
 	frr_with_privs(&zserv_privs) {
-		sd = vrf_socket(PF_INET, SOCK_DGRAM, IPPROTO_IP,
-				interface->vrf->vrf_id, NULL);
+		sd = vrf_socket(PF_INET, SOCK_DGRAM, IPPROTO_IP, vrf_id, NULL);
 		if (sd < 0) {
 			if (IS_ZEBRA_DEBUG_KERNEL)
 				zlog_debug("Failure to read interface %s speed: %d %s",
@@ -292,8 +290,7 @@ static uint32_t get_iflink_speed(struct interface *interface, int *error)
 			return 0;
 		}
 		/* Get the current link state for the interface */
-		rc = vrf_ioctl(interface->vrf->vrf_id, sd, SIOCETHTOOL,
-			       (char *)&ifdata);
+		rc = vrf_ioctl(vrf_id, sd, SIOCETHTOOL, (char *)&ifdata);
 	}
 	if (rc < 0) {
 		if (errno != EOPNOTSUPP && IS_ZEBRA_DEBUG_KERNEL)
@@ -318,11 +315,6 @@ static uint32_t get_iflink_speed(struct interface *interface, int *error)
 	return ret;
 }
 
-uint32_t kernel_get_speed(struct interface *ifp, int *error)
-{
-	return get_iflink_speed(ifp, error);
-}
-
 static ssize_t
 netlink_gre_set_msg_encoder(struct zebra_dplane_ctx *ctx, void *buf,
 			    size_t buflen)
diff --git a/zebra/interface.c b/zebra/interface.c
index f128b5f422..908dd54215 100644
--- a/zebra/interface.c
+++ b/zebra/interface.c
@@ -70,7 +70,7 @@ static void if_zebra_speed_update(struct event *thread)
 
 	zif->speed_checked++;
 
-	new_speed = kernel_get_speed(ifp, &error);
+	new_speed = kernel_get_speed(ifp->vrf->vrf_id, ifp->name, &error);
 
 	/* error may indicate vrf not available or
 	 * interfaces not available.
@@ -2014,7 +2014,7 @@ static void zebra_if_dplane_ifp_handling(struct zebra_dplane_ctx *ctx)
 			if_update_state_mtu6(ifp, mtu);
 			if_update_state_metric(ifp, 0);
 			if (!speed_set)
-				speed = kernel_get_speed(ifp, NULL);
+				speed = kernel_get_speed(ifp->vrf->vrf_id, ifp->name, NULL);
 			if_update_state_speed(ifp, speed);
 			ifp->ptm_status = ZEBRA_PTM_STATUS_UNKNOWN;
 			ifp->txqlen = dplane_ctx_get_intf_txqlen(ctx);
diff --git a/zebra/rt.h b/zebra/rt.h
index e5dc26150a..01104b05a1 100644
--- a/zebra/rt.h
+++ b/zebra/rt.h
@@ -77,7 +77,7 @@ extern int mpls_kernel_init(void);
 void kernel_router_init(void);
 void kernel_router_terminate(void);
 
-extern uint32_t kernel_get_speed(struct interface *ifp, int *error);
+extern uint32_t kernel_get_speed(vrf_id_t vrf_id, const char *ifname, int *error);
 extern int kernel_get_ipmr_sg_stats(struct zebra_vrf *zvrf, void *mroute);
 
 /*
diff --git a/zebra/rt_socket.c b/zebra/rt_socket.c
index de04978600..1e8d3e7eff 100644
--- a/zebra/rt_socket.c
+++ b/zebra/rt_socket.c
@@ -23,6 +23,7 @@
 #include "lib_errors.h"
 
 #include "zebra/debug.h"
+#include "zebra/interface.h"
 #include "zebra/rib.h"
 #include "zebra/rt.h"
 #include "zebra/kernel_socket.h"
@@ -382,9 +383,10 @@ extern int kernel_interface_set_master(struct interface *master,
 	return 0;
 }
 
-uint32_t kernel_get_speed(struct interface *ifp, int *error)
+uint32_t kernel_get_speed(vrf_id_t vrf_id, const char *ifname, int *error)
 {
-	return ifp->speed;
+	*error = INTERFACE_SPEED_ERROR_READ;
+	return 0;
 }
 
 int kernel_upd_mac_nh(uint32_t nh_id, struct in_addr vtep_ip)
-- 
2.43.0

